contracts/casino/include/casino/casino.hpp

```diff
@@ -97,102 +107,123 @@ class [[eosio::contract("casino")]] casino: public eosio::contract {
+        const auto itr = game_state.require_find(game_id, "game not found");
+        game_state.modify(itr, get_self(), [&](auto& row) {
+            row.balance -= quantity;
+        });
+        gstate.game_profits_sum -= quantity;
     }
 
     bool is_active_game(uint64_t game_id) const {
-        game_table games(_self, _self.value);
-        return games.find(game_id) != games.end();
+        return !games.require_find(game_id, "game not found")->paused;
     }
 
     time_point get_last_claim_time(uint64_t game_id) const {
-        game_state_table game_state(_self, _self.value);
-        const auto itr = game_state.find(game_id);
-        return itr == game_state.end() ? time_point() : itr->last_claim_time;
+        const auto itr = game_state.require_find(game_id, "game not found");
+        return itr->last_claim_time;
     }
 
     void update_last_claim_time(uint64_t game_id) {
-        game_state_table game_state(_self, _self.value);
-        const auto itr = game_state.require_find(game_id, "game is not in the casino");
+        const auto itr = game_state.require_find(game_id, "game is not listed in the casino");
         game_state.modify(itr, _self, [&](auto& row) {
             row.last_claim_time = current_time_point();
         });
     }
 
-    void transfer(name to, asset quantity) {
+    void transfer(name to, asset quantity, std::string memo) {
         eosio::action(
             eosio::permission_level{_self, "active"_n},
             "eosio.token"_n,
             "transfer"_n,
-            std::make_tuple(_self, to, quantity, std::string("player winnings"))
+            std::make_tuple(_self, to, quantity, memo)
         ).send();
     }
 
     void verify_game(uint64_t game_id);
     void verify_account(name game_account);
     uint64_t get_game_id(name game_account);
 
-    void session_update(asset quantity) {
+    void session_update(uint64_t game_id, asset quantity) {
+        const auto itr = game_state.require_find(game_id, "game not found");
+        game_state.modify(itr, _self, [&](auto& row) {
+            row.active_sessions_sum += quantity;
+        });
         gstate.game_active_sessions_sum += quantity;
     }
 
-    void session_close(asset quantity) {
+    void session_close(uint64_t game_id, asset quantity) {
+        const auto itr = game_state.require_find(game_id, "game not found");
+        check(quantity <= itr->active_sessions_sum, "invalid quantity in session close");
         check(quantity <= gstate.game_active_sessions_sum, "invalid quantity in session close");
+        check(itr->active_sessions_amount, "no active sessions");
+        check(gstate.active_sessions_amount, "no active sesions");
+
+        game_state.modify(itr, _self, [&](auto& row) {
+            row.active_sessions_sum -= quantity;
+            row.active_sessions_amount--;
+        });
         gstate.game_active_sessions_sum -= quantity;
+        gstate.active_sessions_amount--;
```

Comment:
Should we do `active_sessions_amount--` here, in `session_close`? What about `on_session_close`?

Reply:
either on_session_close or session_close is called. they can't be called simultaneously